ladybug/designday.py

```diff
@@ -1580,21 +1682,33 @@ def radiation_values(self, location, timestep=1):
     def to_dict(self):
         """Convert the Sky Condition to a dictionary."""
         return {
-            'solar_model': self.solar_model,
+            'type': 'ASHRAETau',
             'month': self.month,
             'day_of_month': self.day_of_month,
             'tau_b': self.tau_b,
             'tau_d': self.tau_d,
-            'daylight_savings_indicator': self.daylight_savings_indicator,
-            'type': 'RevisedClearSkyCondition'
+            'daylight_savings': self.daylight_savings
         }
 
+    def __copy__(self):
+        return ASHRAETau(self._month, self._day_of_month, self._tau_b, self._tau_d,
+                         self._daylight_savings)
 
-def _check_analysis_period(analysis_period):
-    assert hasattr(analysis_period, 'isAnalysisPeriod'), 'Expected' \
-        ' AnalysisPeriod type. Got {}'.format(type(analysis_period))
-    assert analysis_period.st_month == analysis_period.end_month and \
-        analysis_period.st_day == analysis_period.end_day and \
-        analysis_period.st_hour == 0 and analysis_period.end_hour == 23, \
-        'analysis_period "{}" does not represent a single day'.format(
-            analysis_period)
+    def __key(self):
+        """A tuple based on the object properties, useful for hashing."""
+        return (self._month, self._day_of_month, self._tau_b, self._tau_d,
+                self._daylight_savings)
+
+    def __hash__(self):
+        return hash(self.__key())
+
+    def __eq__(self, other):
```

Comment:
Two comments:

1. This module is so long. Can we break it down into smaller files? It will be hard to maintain.
2. Do you have a reference for using hash for comparison methods?

Reply:
Also it might be a good idea to cache the hash when possible. Specially for immutable objects. Here is how datetime module does it:

https://github.com/python/cpython/blob/96b1c59c71534db3f0f3799cd84e2006923a5098/Lib/datetime.py#L770-L772

Reply:
I agree that is a good idea but `DesignDays` and `Locations` are mutable and I think they really need to be so this suggestion isn't applicable here. Furthermore, most of the objects in `honeybee_energy` are mutable but they are also lockable. It's possible that we can cache the hash while it's locked but I wonder if all the acrobatics that are necessary to reset this cache when it's unlocked will negate the performance savings of caching. Really, the only place in our libraries where this is definitely applicable is `ladybug_geometry` since all geometry objects are immutable. But those don't possess any `__hash__` or `__eq__` functionality right now. It's probably a good idea to add this, though so I opened an issue and I made a note about caching the hash there: 
https://github.com/ladybug-tools/ladybug-geometry/issues/80

Reply:
My main question about the idea of using hash for comparison remains. How did you find this to be a good practice? Did you find this implementation in one of the standard libraries or did you read about it somewhere?

Reply:
I don't use `__hash__` for comparison. `__hash__` only gets used by the native Python `set()` and `hash()` operators (not `==` or `!=`).

It's true that I use a `__key` for all 4 of these operators, though, so that there's only one place where all of the unique properties of the object are defined. This way, I don't have to worry about a case where `set()` evaluates two different objects but `==` tells me that the two objects are equivalent.  This particular suggestion comes from the most up-voted suggestion on StackOverflow:

https://stackoverflow.com/questions/2909106/whats-a-correct-and-good-way-to-implement-hash

Reply:
Thanks but the example is different from your implementation. In the example the `__key` is a tuple of attributes. In your case you have nested hashing for each item in the tuple. Is that necessary?

Reply:
I see your question now. Yes, it's necessary to hash it because nested tuples cannot be hashed. By hashing them within the `__key`, we are able to ensure everything is hash-able up the whole nested hierarchy.

Reply:
I researched this a bit more and it turns out that [hash collisions](https://preshing.com/20110504/hash-collision-probabilities/) are, in fact, a thing that could happen. But I think it follows from logic that the probability of encountering these with the nested hashes here isn't any larger than the probability of encountering such a collision with a non-nested hash. And I have some faith that the hashing function under the hood of modern Python installations has a low enough probability of collision that we won't encounter it (hopefully, those are not famous last words!).